core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -412,6 +432,11 @@ private[spark] class Executor(
           env.mapOutputTracker.asInstanceOf[MapOutputTrackerWorker].updateEpoch(task.epoch)
         }
 
+        val stageId = task.stageId
+        val stageAttemptId = task.stageAttemptId
+        metricsPoller.onTaskStart(taskId, stageId, stageAttemptId)
+        taskStarted = true
```

Comment:
This boolean is needed because we call metricsPoller.onTaskStart in this try block here and in the finally block we want to call metricsPoller.onTaskCompletion to do cleanup. However, we do not want to call metricsPoller.onTaskCompletion if we exited the try block before metricsPoller.onTaskStart has been called. The contract for ExecutorMetricsPoller is that onTaskCompletion should only be called if onTaskStart has been called first with the same arguments.
There is a comment in the finally block explaining the significance taskStarted. I didn't think I need a comment here as well. But let me see if I can add a suitable comment before the finally block.

Reply:
Sorry my comment was not clear enough. So if I recall correctly what I meant here is the method `onTaskCompletion()` seamed to me handling this case (when the task is not started) very well:

https://github.com/apache/spark/blob/f9c284bdc650e2c06d06b5488d637ffad3390d99/core/src/main/scala/org/apache/spark/executor/ExecutorMetricsPoller.scala#L148-L151

So I would save the extra unnecessary boolean flag here where the code is already quite complex and long. But if you want to make this part robust for future code changes then you can be more explicit within `onTaskCompletion()` by adding an `if` wrapping these two lines and checking the content of `stageTCMP`. 

What is your opinion?

Reply:
If the code here is complex and long, it is because it was already so before my change. As a matter of fact, I added relatively few lines of code to the method here.
The taskStarted flag actually signals two things; I only mentioned one previously, that metricsPoller.onTaskStart has been called. It also signals that the task has been deserialized, since that has to happen first. Therefore, in the finally block, if taskStarted is true, it is safe to use task.stageId and task.stageAttemptId, and we them in order to call metricsPoller.onTaskCompletion. It is not safe to simply call metricsPoller.onTaskCompletion in the finally block without any check. The taskStarted flag is a convenient check.
Thus I'd argue that it is necessary. Sure, we can replace it with a different check but that wouldn't be simpler. In my view, the taskStarted flag is self-explanatory and simple to understand.

Reply:
Ok. Basically both solutions are working:
- having `taskStarted` as an extra `var`
- using only the existing `task != null` in the `finally` block

I still prefer the second (as it is reusing the existing `var`) but I can accept the first solution too.